### DISCUSSION TITLE
Deploy apps on subdomain

### DISCUSSION THREAD

**Discussion by giannoug** [onedr0p/cluster-template#1034] [Q&A] [answered]
Is it possible to deploy multiple clusters under the same Cloudflare domain? I just gave it a try using `bootstrap_cloudflare_domain: rpi4.example.com` and even though it completed successfully, CF domains are not respecting that and are using the base domain.

The idea behind that is to use one base domain for a couple of clusters, testing, etc.

**Comment by onedr0p** [comment]
~~The template doesn't support this out of the box since it uses a wildcart cert for the base domain. cloudflare doesn't support wildcard certs for sub-domains unless you pay. You should be able to get around it thou by not using wildcard certs and instead having cert-manager create a cert per sub-domain using a cert-manager cluster issuer ingress annotation.~~

That was wrong I was thinking of a wildcard DNS off the base domain, hopefully I can follow up with a correct answer lol

I converted this issue to a discussion.

**Reply by onedr0p** [reply]
I am having trouble understanding:
> CF domains are not respecting that and are using the base domain.

can you expand on that?

**Reply by giannoug** [reply]
Sure! I chose the wrong words there. What I meant is that DNS records created do not respect `bootstrap_cloudflare_domain` (which in my original post was `rpi4.example.com`) and seem to use `external.example.com` directly. I expect a CNAME record with the value of `external.rpi4.example.com`

![image](https://github.com/onedr0p/flux-cluster-template/assets/509474/e4a35af1-ee01-4d95-9896-6eba37863751)

**Reply by onedr0p** [reply]
I am not entirely sure why that is happening, but I cant point you to external-dns which is creating those CNAMEs off of the `external` CNAME. Maybe try to enable verbose logging with external-dns and check the logs on it.

The `external` CNAME is created here https://github.com/onedr0p/flux-cluster-template/blob/main/bootstrap/templates/kubernetes/apps/networking/cloudflared/app/dnsendpoint.yaml.j2

I am curious if you find out anything so hopefully it can be supported or at least solved for anyone else trying this. I think most people just use different domains since they are cheap anyways to buy.

**Reply by giannoug** [reply]
Hmm, that looks correct. I'll try from scratch, check again and report back

**Reply by giannoug** [reply]
I nuked the cluster, deleted all records from cloudflare, created new repo from scratch and followed the setup steps. I can't see any records being created now, apart from a TXT one that eventually got deleted (I guess for SSL verification -- meaning that the keys are correct).

**Reply by giannoug** [reply]
`kubectl get ingress -A` shows the correct domains though!

**Reply by onedr0p** [reply]
If the records aren't getting created in CF I would start looking around external-dns logs. You might have to enable debug logging which you can do via setting debug on https://github.com/kubernetes-sigs/external-dns/blob/master/charts/external-dns/values.yaml#L178

**Reply by giannoug** [reply]
Here's what I've found:
```
time="2023-12-09T23:51:45Z" level=info msg="config: {APIServerURL: KubeConfig: RequestTimeout:30s DefaultTargets:[] GlooNamespace:gloo-system SkipperRouteGroupVersion:zalando.org/v1 Sources:[crd ingress] Namespace: AnnotationFilter: LabelFilter: IngressClassNames:[external] FQDNTemplate: CombineFQDNAndAnnotation:false IgnoreHostnameAnnotation:false IgnoreIngressTLSSpec:false IgnoreIngressRulesSpec:false GatewayNamespace: GatewayLabelFilter: Compatibility: PublishInternal:false PublishHostIP:false AlwaysPublishNotReadyAddresses:false ConnectorSourceServer:localhost:8080 Provider:cloudflare GoogleProject: GoogleBatchChangeSize:1000 GoogleBatchChangeInterval:1s GoogleZoneVisibility: DomainFilter:[subdomain.example.net] ExcludeDomains:[] RegexDomainFilter: RegexDomainExclusion: ZoneNameFilter:[] ZoneIDFilter:[] TargetNetFilter:[] ExcludeTargetNets:[] AlibabaCloudConfigFile:/etc/kubernetes/alibaba-cloud.json AlibabaCloudZoneType: AWSZoneType: AWSZoneTagFilter:[] AWSAssumeRole: AWSAssumeRoleExternalID: AWSBatchChangeSize:1000 AWSBatchChangeInterval:1s AWSEvaluateTargetHealth:true AWSAPIRetries:3 AWSPreferCNAME:false AWSZoneCacheDuration:0s AWSSDServiceCleanup:false AWSDynamoDBRegion: AWSDynamoDBTable:external-dns AzureConfigFile:/etc/kubernetes/azure.json AzureResourceGroup: AzureSubscriptionID: AzureUserAssignedIdentityClientID: BluecatDNSConfiguration: BluecatConfigFile:/etc/kubernetes/bluecat.json BluecatDNSView: BluecatGatewayHost: BluecatRootZone: BluecatDNSServerName: BluecatDNSDeployType:no-deploy BluecatSkipTLSVerify:false CloudflareProxied:true CloudflareDNSRecordsPerPage:100 CoreDNSPrefix:/skydns/ RcodezeroTXTEncrypt:false AkamaiServiceConsumerDomain: AkamaiClientToken: AkamaiClientSecret: AkamaiAccessToken: AkamaiEdgercPath: AkamaiEdgercSection: InfobloxGridHost: InfobloxWapiPort:443 InfobloxWapiUsername:admin InfobloxWapiPassword: InfobloxWapiVersion:2.3.1 InfobloxSSLVerify:true InfobloxView: InfobloxMaxResults:0 InfobloxFQDNRegEx: InfobloxNameRegEx: InfobloxCreatePTR:false InfobloxCacheDuration:0 DynCustomerName: DynUsername: DynPassword: DynMinTTLSeconds:0 OCIConfigFile:/etc/kubernetes/oci.yaml OCICompartmentOCID: OCIAuthInstancePrincipal:false InMemoryZones:[] OVHEndpoint:ovh-eu OVHApiRateLimit:20 PDNSServer:http://localhost:8081 PDNSAPIKey: PDNSSkipTLSVerify:false TLSCA: TLSClientCert: TLSClientCertKey: Policy:sync Registry:txt TXTOwnerID:default TXTPrefix:k8s. TXTSuffix: TXTEncryptEnabled:false TXTEncryptAESKey: Interval:1m0s MinEventSyncInterval:5s Once:false DryRun:false UpdateEvents:false LogFormat:text MetricsAddress::7979 LogLevel:debug TXTCacheInterval:0s TXTWildcardReplacement: ExoscaleEndpoint: ExoscaleAPIKey: ExoscaleAPISecret: ExoscaleAPIEnvironment:api ExoscaleAPIZone:ch-gva-2 CRDSourceAPIVersion:externaldns.k8s.io/v1alpha1 CRDSourceKind:DNSEndpoint ServiceTypeFilter:[] CFAPIEndpoint: CFUsername: CFPassword: ResolveServiceLoadBalancerHostname:false RFC2136Host: RFC2136Port:0 RFC2136Zone: RFC2136Insecure:false RFC2136GSSTSIG:false RFC2136KerberosRealm: RFC2136KerberosUsername: RFC2136KerberosPassword: RFC2136TSIGKeyName: RFC2136TSIGSecret: RFC2136TSIGSecretAlg: RFC2136TAXFR:false RFC2136MinTTL:0s RFC2136BatchChangeSize:50 NS1Endpoint: NS1IgnoreSSL:false NS1MinTTLSeconds:0 TransIPAccountName: TransIPPrivateKeyFile: DigitalOceanAPIPageSize:50 ManagedDNSRecordTypes:[A AAAA CNAME] GoDaddyAPIKey: GoDaddySecretKey: GoDaddyTTL:0 GoDaddyOTE:false OCPRouterName: IBMCloudProxied:false IBMCloudConfigFile:/etc/kubernetes/ibmcloud.json TencentCloudConfigFile:/etc/kubernetes/tencent-cloud.json TencentCloudZoneType: PiholeServer: PiholePassword: PiholeTLSInsecureSkipVerify:false PluralCluster: PluralProvider:}"
time="2023-12-09T23:51:45Z" level=info msg="Instantiating new Kubernetes client"
time="2023-12-09T23:51:45Z" level=debug msg="apiServerURL: "
time="2023-12-09T23:51:45Z" level=debug msg="kubeConfig: "
time="2023-12-09T23:51:45Z" level=info msg="Using inCluster-config based on serviceaccount-token"
time="2023-12-09T23:51:45Z" level=info msg="Created Kubernetes client https://10.43.0.1:443"
time="2023-12-09T23:51:45Z" level=debug msg="no zoneIDFilter configured, looking at all zones"
time="2023-12-09T23:51:46Z" level=debug msg="zone example.net not in domain filter"
time="2023-12-09T23:51:46Z" level=debug msg="Discarding ingress kube-system/hubble-ui because it does not match required ingress classes [external]"
time="2023-12-09T23:51:46Z" level=debug msg="Endpoints generated from ingress: flux-system/flux-webhook: [flux-webhook.subdomain.example.net 0 IN CNAME  external.subdomain.example.net [] flux-webhook.subdomain.example.net 0 IN CNAME  external.subdomain.example.net []]"
time="2023-12-09T23:51:46Z" level=debug msg="Endpoints generated from ingress: networking/echo-server: [echo-server.subdomain.example.net 0 IN CNAME  external.subdomain.example.net [] echo-server.subdomain.example.net 0 IN CNAME  external.subdomain.example.net []]"
time="2023-12-09T23:51:46Z" level=debug msg="Removing duplicate endpoint flux-webhook.subdomain.example.net 0 IN CNAME  external.subdomain.example.net []"
time="2023-12-09T23:51:46Z" level=debug msg="Removing duplicate endpoint echo-server.subdomain.example.net 0 IN CNAME  external.subdomain.example.net []"
time="2023-12-09T23:51:46Z" level=debug msg="no zoneIDFilter configured, looking at all zones"
time="2023-12-09T23:51:47Z" level=debug msg="zone example.net not in domain filter"
time="2023-12-09T23:51:47Z" level=debug msg="Skipping record echo-server.subdomain.example.net because no hosted zone matching record DNS Name was detected"
time="2023-12-09T23:51:47Z" level=debug msg="Skipping record external.subdomain.example.net because no hosted zone matching record DNS Name was detected"
time="2023-12-09T23:51:47Z" level=debug msg="Skipping record flux-webhook.subdomain.example.net because no hosted zone matching record DNS Name was detected"
time="2023-12-09T23:51:47Z" level=debug msg="Skipping record k8s.echo-server.subdomain.example.net because no hosted zone matching record DNS Name was detected"
time="2023-12-09T23:51:47Z" level=debug msg="Skipping record k8s.cname-echo-server.subdomain.example.net because no hosted zone matching record DNS Name was detected"
time="2023-12-09T23:51:47Z" level=debug msg="Skipping record k8s.external.subdomain.example.net because no hosted zone matching record DNS Name was detected"
time="2023-12-09T23:51:47Z" level=debug msg="Skipping record k8s.cname-external.subdomain.example.net because no hosted zone matching record DNS Name was detected"
time="2023-12-09T23:51:47Z" level=debug msg="Skipping record k8s.flux-webhook.subdomain.example.net because no hosted zone matching record DNS Name was detected"
time="2023-12-09T23:51:47Z" level=debug msg="Skipping record k8s.cname-flux-webhook.subdomain.example.net because no hosted zone matching record DNS Name was detected"
```

**Reply by onedr0p** [reply]
I google that error `skipping record because no hosted zone matching record DNS Name was detected` and found [this](https://github.com/kubernetes-sigs/external-dns/issues/304).

My guess is that `${SECRET_DOMAIN}` in the external-dns helmrelease needs to be `example.net`, not `subdomain.example.net`

Try updating that value [here](https://github.com/onedr0p/flux-cluster-template/blob/e14d5f50716bc0146001a309672bb985f6a628d3/bootstrap/templates/kubernetes/apps/networking/external-dns/app/helmrelease.yaml.j2#L44) to be your base domain, not with the subdomain. You can try to use a [fancy function](https://github.com/drone/envsubst#supported-functions) to strip off the subdomain maybe instead of creating a new var and if you get that to work let me know 😄

**Reply by giannoug** [reply]
I patched the deployment to see if it's working and it seems that it does, at least partially! Records are created, but CF has some warnings (not sure if that's normal or not):
![image](https://github.com/onedr0p/flux-cluster-template/assets/509474/a81f7250-f72c-488e-b752-e267d72d44a9)
Also Chrome doesn't seem to like something with the certs because I'm getting `ERR_SSL_VERSION_OR_CIPHER_MISMATCH`.

Is it due to letsencrypt-staging issuer?

**Reply by onedr0p** [reply]
`ERR_SSL_VERSION_OR_CIPHER_MISMATCH` could be related to using `letsencrypt-staging`. However I am more concerned with that error on Cloudflares side. What is in the green? It looks too short to be the domain, is it only the subdomain?

This is what mine looks like without a subdomain.domain.tld

<img width="804" alt="image" src="https://github.com/onedr0p/flux-cluster-template/assets/213795/a8abee25-e6a7-4cc1-a5b7-51ba5524d124">

**Reply by giannoug** [reply]
Could be the same error, according to this: https://github.com/cloudflare/cloudflare-docs/blob/production/content/ssl/edge-certificates/additional-options/total-tls/error-messages.md#error-messages

It's only the subdomain, yes. Records are created correctly, I can resolve them from my machine :tada: I'll enable the production issuer and report back

**Reply by giannoug** [reply]
Hmm, no luck. New cert got installed, at least according to `task cluster:certificates` but I'm still getting ERR_SSL_VERSION_OR_CIPHER_MISMATCH.

**Reply by onedr0p** [reply]
Without trying this myself I am unsure how to guide you into fixing this issue. I hope you can poke around and get it working though! My only suggestion in the meantime would be buy a new domain (they are cheap) and use that instead. I imagine this isn't the only issue you will run into :/

**Reply by giannoug** [reply]
Yeah I agree at this point. Money isn't the issue, I thought I could group servers under the same domain. I'm going to get some new ones for the time being. Thanks for your help!

**Comment by onedr0p** [comment, marked as answer]
To circle back on this multi-level subdomains like you want to do here are not covered by the CF Universal Certificate

https://developers.cloudflare.com/ssl/troubleshooting/version-cipher-mismatch/#multi-level-subdomains
